File: src/student_execution_os/web/static/js/commands.js
// Commands about things the user already has, in everyday RU/EN phrasing — the
// device twin of agent/commands.py (same grammar, same fixture:
// tests/fixtures/nl_command_cases.json), so "готово эссе" or "перенеси созвон на
// 18:00" also work offline. A parse is only a proposal: it is previewed, and
// closing/archiving anything is confirmed before a sync operation is queued.
import { parseTask } from './nlparse.js';

const PUNCT = /["'«»“”„()!?.,;:]+/gu;
const LEAD_PREPS = /^(?:над|по|на|про|о|об|с|со|задачу|задача|событие|напоминание|on|for|about|of|at|the|task|event|reminder)\s+/iu;
const CLOCK_WORDS = /\d{1,2}[:.]\d{2}|(?<!\p{L})(?:в|во|к|at|by)\s+\d{1,2}(?!\d)|\d\s*(?:am|pm)|утр|вечер|дн[её]м|ноч|полдень|полночь|morning|evening|afternoon|night|noon|midnight|через|in\s+\d/iu;

const COMPLETE = /^(?:готово|сделано|сделал[аи]?|выполнил[аи]?|выполнено|закончил[аи]?|отметь(?:\s+как)?\s+выполненн\p{L}*|done|complete|completed|finished|mark\s+done)\s*[-—:]?\s+(?<target>.+)$/iu;
const COMPLETE_SUFFIX = /^(?<target>.+?)\s*[-—:]?\s+(?:готово|сделано|выполнено|done)$/iu;
const ARCHIVE = /^(?:в\s+архив|архивируй|заархивируй|убери\s+в\s+архив|archive)\s*[-—:]?\s+(?<target>.+)$/iu;
const ARCHIVE_SUFFIX = /^(?<target>.+?)\s+в\s+архив$/iu;
const CANCEL = /^(?:отмени(?:ть)?|не\s+буду\s+делать|cancel|drop)\s*[-—:]?\s+(?<target>.+)$/iu;
const MOVE = /^(?:перенеси(?:те)?|перенести|сдвинь|передвинь|move|reschedule|postpone)\s+(?<rest>.+)$/iu;
const SNOOZE = /^(?:напомни(?:те)?(?:\s+мне)?\s+(?:про|о|об)|remind\s+me\s+(?:about|of)|отложи|snooze)\s+(?<rest>.+)$/iu;
const PROGRESS = /^(?:поработал[аи]?|позанимал(?:ся|ась|ись)|занимал(?:ся|ась|ись)|потратил[аи]?|worked|spent|logged)\s+(?<rest>.+)$/iu;
const COUNT = /^(?:сделал[аи]?|решил[аи]?|прочитал[аи]?|написал[аи]?|did|solved|read|finished)\s+(?<n>\d{1,5})\s+(?<unit>\S+)\s+(?:по|из|в|для|for|of|in|from)\s+(?<target>.+)$/iu;

export function normalize(text) {
  return String(text || '').toLowerCase().replaceAll('ё', 'е').replace(PUNCT, ' ').split(/\s+/u).filter(Boolean).join(' ');
}

// A crude RU/EN stem: endings are vowels ("лабу"/"лаба", "матану"/"матан").
const stem = (word) => (word.replace(/[аеиоуыэюяйьeiouy]+$/u, '') || word).slice(0, 5);

function stems(text) {
  return new Set(normalize(text).split(' ').filter((w) => w.length > 2 || /^\d+$/u.test(w)).map(stem));
}

// How well a spoken fragment names an item (1.0 exact … 0 unrelated).
export function score(fragment, title) {
  const a = normalize(fragment);
  const b = normalize(title);
  if (!a || !b) return 0;
  if (a === b) return 1;
  if (a.length >= 3 && (b.includes(a) || a.includes(b))) return 0.9;
  const sa = stems(a);
  const sb = stems(b);
  if (!sa.size || !sb.size) return 0;
  if ([...sa].every((x) => sb.has(x))) return 0.85;
  const common = [...sa].filter((x) => sb.has(x)).length;
  return common / new Set([...sa, ...sb]).size;
}

// The single item the fragment names, or null plus the closest candidates.
export function matchTarget(fragment, items, { kinds = ['TASK', 'EVENT', 'REMINDER'], statuses = null } = {}) {
  const pool = (items || []).filter((x) => kinds.includes(x.kind) && (!statuses || statuses.includes(x.status)));
  const scored = pool.map((item, index) => ({ value: score(fragment, item.title), index, item }))
    .sort((a, b) => b.value - a.value || a.index - b.index);
  const candidates = scored.filter((x) => x.value >= 0.34).slice(0, 5).map((x) => x.item);
  if (!scored.length || scored[0].value < 0.5) return { item: null, candidates };
  if (scored.length > 1 && scored[1].value >= scored[0].value - 0.05) return { item: null, candidates };
  return { item: scored[0].item, candidates };
}

function cleanTarget(text) {
  let value = String(text).split(/\s+/u).filter(Boolean).join(' ').replace(/^[\s\-—:,.]+|[\s\-—:,.]+$/gu, '');
  for (;;) {
    const stripped = value.replace(LEAD_PREPS, '');
    if (stripped === value) return value;
    value = stripped;
  }
}

function momentOf(parsed) {
  for (const key of ['starts_at', 'remind_at', 'actionable_from', 'target_at']) if (parsed[key]) return new Date(parsed[key]);
  return parsed.actual_cutoff?.at ? new Date(parsed.actual_cutoff.at) : null;
}

function action(command, payload, { item, target, confirm = false, unresolved = [] }) {
  const missing = [...unresolved];
  let body = payload;
  if (!item) { body = { ...payload, target_text: target }; missing.unshift('target'); }
  else body = { [item.kind === 'REMINDER' ? 'reminder_id' : 'obligation_id']: item.id, ...payload };
  return { command, payload: body, confidence: item ? 0.9 : 0.5, unresolved_fields: missing,
    expected_version: item ? item.version ?? null : null, requires_confirmation: confirm };
}

// The sync operation that moves a task, event or reminder to `when`. `keepTime`:
// only a day was said, so the item keeps its own time of day.
export function rescheduleChange(kind, entity, when, keepTime) {
  let current;
  if (kind === 'EVENT') current = entity.starts_at;
  else if (kind === 'REMINDER') current = entity.remind_at;
  else current = entity.actual_cutoff?.state === 'KNOWN' ? entity.actual_cutoff.at : entity.actionable_from;
  let at = new Date(when);
  if (keepTime && current) {
    const old = new Date(current);
    at = new Date(at.getFullYear(), at.getMonth(), at.getDate(), old.getHours(), old.getMinutes());
  }
  const value = at.toISOString();
  if (kind === 'EVENT') return ['event.update', { starts_at: value }];
  if (kind === 'REMINDER') return ['reminder.update', { remind_at: value }];
  if (entity.actual_cutoff?.state === 'KNOWN') return ['task.update', { actual_cutoff: { state: 'KNOWN', at: value, boundary: entity.actual_cutoff.boundary || 'INCLUSIVE' } }];
  return ['task.defer', { until: value }];
}

// parseCommand(text, now, items) → one typed action, or null (read it as something new).
// items: [{ id, kind: TASK|EVENT|REMINDER, title, status, version }]
export function parseCommand(text, now = new Date(), items = []) {
  const raw = String(text || '').split(/\s+/u).filter(Boolean).join(' ').replace(/^[\s.!]+|[\s.!]+$/gu, '');
  if (!raw) return null;
  const open = ['ACTIVE', 'DRAFT', 'SCHEDULED', 'FIRED'];

  let m = COUNT.exec(raw);
  if (m) {
    const target = cleanTarget(m.groups.target);
    const { item } = matchTarget(target, items, { kinds: ['TASK'], statuses: ['ACTIVE', 'DRAFT'] });
    if (item || !items.length) return action('LOG_PROGRESS', { count: Number(m.groups.n) }, { item, target });
  }
  m = PROGRESS.exec(raw);
  if (m) {
    const parsed = parseTask(m.groups.rest, now);
    const minutes = parsed.estimated_total_effort_minutes;
    const target = cleanTarget(parsed.title || '');
    if (minutes && target) {
      const { item } = matchTarget(target, items, { kinds: ['TASK'], statuses: ['ACTIVE', 'DRAFT'] });
      return action('LOG_PROGRESS', { minutes: Number(minutes) }, { item, target });
    }
  }
  for (const [pattern, command, kinds, statuses, suffix] of [
    [COMPLETE, 'COMPLETE_OBLIGATION', ['TASK', 'REMINDER'], open, false],
    [ARCHIVE, 'ARCHIVE_OBLIGATION', ['TASK'], null, false],
    [CANCEL, 'CANCEL_OBLIGATION', ['TASK', 'EVENT', 'REMINDER'], open, false],
    [COMPLETE_SUFFIX, 'COMPLETE_OBLIGATION', ['TASK', 'REMINDER'], open, true],
    [ARCHIVE_SUFFIX, 'ARCHIVE_OBLIGATION', ['TASK'], null, true],
  ]) {
    const hit = pattern.exec(raw);
    if (!hit) continue;
    const target = cleanTarget(hit.groups.target);
    const { item } = matchTarget(target, items, { kinds, statuses });
    if (!item && suffix) continue; // only a known item makes a suffix a command
    return action(command, {}, { item, target, confirm: true });
  }
  m = MOVE.exec(raw);
  if (m) {
    const parsed = parseTask(m.groups.rest, now);
    const when = momentOf(parsed);
    const target = cleanTarget(parsed.title || m.groups.rest);
    const { item } = matchTarget(target, items, { statuses: open });
    const payload = when ? { when: when.toISOString(), keep_time: !CLOCK_WORDS.test(m.groups.rest) } : {};
    return action('RESCHEDULE', payload, { item, target, unresolved: when ? [] : ['when'] });
  }
  m = SNOOZE.exec(raw);
  if (m) {
    const parsed = parseTask(m.groups.rest, now);
    let until = momentOf(parsed);
    if (!until && parsed.estimated_total_effort_minutes) until = new Date(now.getTime() + parsed.estimated_total_effort_minutes * 60000);
    const target = cleanTarget(parsed.title || m.groups.rest);
    const { item } = matchTarget(target, items, { kinds: ['TASK', 'REMINDER'], statuses: open });
    if (!item || !until) return null; // "напомни про встречу с деканом завтра": a new reminder
    return action('SNOOZE', { until: until.toISOString() }, { item, target });
  }
  return null;
}

File: src/student_execution_os/agent/commands.py
"""Commands about things the user already has, in everyday RU/EN phrasing.

    "готово эссе", "эссе в архив", "отмени созвон", "перенеси эссе на завтра в 18",
    "напомни про эссе через час", "поработал над эссе 30 минут", "сделал 3 задачи по матану"

``parse_command`` turns such a phrase into one typed Assistant action (the same
schema a language model must produce) or returns ``None`` so the text is read as
something new to create. It never executes anything: every action is previewed and,
for destructive ones, confirmed by the user.

The device runs an identical grammar (web/static/js/commands.js) against the same
fixture (tests/fixtures/nl_command_cases.json), so commands also work offline.
"""
from __future__ import annotations

import re
from datetime import datetime, timedelta
from typing import Any
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from .nlparse import parse_task

_PUNCT = re.compile(r"[\"'«»“”„()!?.,;:]+")
_LEAD_PREPS = re.compile(r"^(?:над|по|на|про|о|об|с|со|задачу|задача|событие|напоминание|on|for|about|of|at|the|task|event|reminder)\s+", re.I)
_CLOCK_WORDS = re.compile(r"\d{1,2}[:.]\d{2}|(?<!\w)(?:в|во|к|at|by)\s+\d{1,2}(?!\d)|\d\s*(?:am|pm)|утр|вечер|дн[её]м|ноч|полдень|полночь"
                          r"|morning|evening|afternoon|night|noon|midnight|через|in\s+\d", re.I)

COMPLETE = re.compile(r"^(?:готово|сделано|сделал[аи]?|выполнил[аи]?|выполнено|закончил[аи]?|отметь(?:\s+как)?\s+выполненн\w*"
                      r"|done|complete|completed|finished|mark\s+done)\s*[-—:]?\s+(?P<target>.+)$", re.I)
COMPLETE_SUFFIX = re.compile(r"^(?P<target>.+?)\s*[-—:]?\s+(?:готово|сделано|выполнено|done)$", re.I)
ARCHIVE = re.compile(r"^(?:в\s+архив|архивируй|заархивируй|убери\s+в\s+архив|archive)\s*[-—:]?\s+(?P<target>.+)$", re.I)
ARCHIVE_SUFFIX = re.compile(r"^(?P<target>.+?)\s+в\s+архив$", re.I)
CANCEL = re.compile(r"^(?:отмени(?:ть)?|не\s+буду\s+делать|cancel|drop)\s*[-—:]?\s+(?P<target>.+)$", re.I)
MOVE = re.compile(r"^(?:перенеси(?:те)?|перенести|сдвинь|передвинь|move|reschedule|postpone)\s+(?P<rest>.+)$", re.I)
SNOOZE = re.compile(r"^(?:напомни(?:те)?(?:\s+мне)?\s+(?:про|о|об)|remind\s+me\s+(?:about|of)|отложи|snooze)\s+(?P<rest>.+)$", re.I)
PROGRESS = re.compile(r"^(?:поработал[аи]?|позанимал(?:ся|ась|ись)|занимал(?:ся|ась|ись)|потратил[аи]?|worked|spent|logged)\s+(?P<rest>.+)$", re.I)
COUNT = re.compile(r"^(?:сделал[аи]?|решил[аи]?|прочитал[аи]?|написал[аи]?|did|solved|read|finished)\s+(?P<n>\d{1,5})\s+(?P<unit>\S+)"
                   r"\s+(?:по|из|в|для|for|of|in|from)\s+(?P<target>.+)$", re.I)


def normalize(text: str) -> str:
    return " ".join(_PUNCT.sub(" ", str(text or "").casefold().replace("ё", "е")).split())


def _stem(word: str) -> str:
    """A crude RU/EN stem: endings are vowels ("лабу"/"лаба", "матану"/"матан")."""
    return (word.rstrip("аеиоуыэюяйьeiouy") or word)[:5]


def _stems(text: str) -> set[str]:
    return {_stem(word) for word in normalize(text).split() if len(word) > 2 or word.isdigit()}


def score(fragment: str, title: str) -> float:
    """How well a spoken fragment names an item (1.0 exact … 0 unrelated)."""
    a, b = normalize(fragment), normalize(title)
    if not a or not b:
        return 0.0
    if a == b:
        return 1.0
    if len(a) >= 3 and (a in b or b in a):
        return 0.9
    sa, sb = _stems(a), _stems(b)
    if not sa or not sb:
        return 0.0
    if sa <= sb:
        return 0.85
    return len(sa & sb) / len(sa | sb)


def match_target(fragment: str, items: list[dict[str, Any]], *, kinds: tuple[str, ...] = ("TASK", "EVENT", "REMINDER"),
                 statuses: tuple[str, ...] | None = None) -> tuple[dict[str, Any] | None, list[dict[str, Any]]]:
    """The single item the fragment names, or None plus the closest candidates."""
    pool = [item for item in items if item.get("kind") in kinds and (statuses is None or item.get("status") in statuses)]
    scored = sorted(((score(fragment, str(item.get("title", ""))), index, item) for index, item in enumerate(pool)),
                    key=lambda row: (-row[0], row[1]))
    candidates = [item for value, _, item in scored if value >= 0.34][:5]
    if not scored or scored[0][0] < 0.5:
        return None, candidates
    if len(scored) > 1 and scored[1][0] >= scored[0][0] - 0.05:
        return None, candidates  # two items fit equally well: the user chooses
    return scored[0][2], candidates


def _clean_target(text: str) -> str:
    value = " ".join(str(text).split()).strip(" -—:,.")
    while True:
        stripped = _LEAD_PREPS.sub("", value)
        if stripped == value:
            return value
        value = stripped


def _moment(parsed: dict[str, Any]) -> datetime | None:
    for key in ("starts_at", "remind_at", "actionable_from", "target_at"):
        if parsed.get(key):
            return datetime.fromisoformat(parsed[key])
    cutoff = parsed.get("actual_cutoff") or {}
    return datetime.fromisoformat(cutoff["at"]) if cutoff.get("at") else None


def _action(command: str, payload: dict[str, Any], *, item: dict[str, Any] | None, target: str,
            confirm: bool = False, unresolved: list[str] | None = None) -> dict[str, Any]:
    missing = list(unresolved or [])
    if item is None:
        payload = {**payload, "target_text": target}
        missing = ["target", *missing]
    else:
        payload = {"reminder_id" if item["kind"] == "REMINDER" else "obligation_id": item["id"], **payload}
    return {"command": command, "payload": payload, "confidence": 0.9 if item else 0.5,
            "unresolved_fields": missing, "expected_version": None if item is None else item.get("version"),
            "requires_confirmation": confirm}


def reschedule_change(kind: str, entity: dict[str, Any], when: datetime, keep_time: bool,
                      zone: ZoneInfo) -> tuple[str, dict[str, Any]]:
    """The sync operation that moves a task, event or reminder to ``when``.

    ``keep_time``: only a day was said ("перенеси на завтра"), so the item keeps its
    own time of day. A task with a deadline moves its deadline; one without is put
    off until then; an event moves its start (its length is kept); a reminder moves
    its moment.
    """
    if kind == "EVENT":
        current = entity.get("starts_at")
    elif kind == "REMINDER":
        current = entity.get("remind_at")
    else:
        cutoff = entity.get("actual_cutoff") or {}
        current = cutoff.get("at") if cutoff.get("state") == "KNOWN" else entity.get("actionable_from")
    if keep_time and current:
        old = datetime.fromisoformat(current).astimezone(zone)
        when = datetime.combine(when.astimezone(zone).date(), old.time(), zone)
    value = when.astimezone(ZoneInfo("UTC")).isoformat()
    if kind == "EVENT":
        return "event.update", {"starts_at": value}
    if kind == "REMINDER":
        return "reminder.update", {"remind_at": value}
    cutoff = entity.get("actual_cutoff") or {}
    if cutoff.get("state") == "KNOWN":
        return "task.update", {"actual_cutoff": {"state": "KNOWN", "at": value, "boundary": cutoff.get("boundary") or "INCLUSIVE"}}
    return "task.defer", {"until": value}


def parse_command(text: str, *, now: datetime, timezone_name: str, items: list[dict[str, Any]]) -> dict[str, Any] | None:
    raw = " ".join(str(text or "").split()).strip(" .!")
    if not raw:
        return None
    try:
        zone = ZoneInfo(timezone_name or "UTC")
    except (ZoneInfoNotFoundError, ValueError):
        zone = ZoneInfo("UTC")
    open_ = ("ACTIVE", "DRAFT", "SCHEDULED", "FIRED")

    counted = COUNT.match(raw)
    if counted:
        target = _clean_target(counted.group("target"))
        item, _ = match_target(target, items, kinds=("TASK",), statuses=("ACTIVE", "DRAFT"))
        if item is not None or not items:
            return _action("LOG_PROGRESS", {"count": int(counted.group("n"))}, item=item, target=target)
    worked = PROGRESS.match(raw)
    if worked:
        parsed = parse_task(worked.group("rest"), now=now, timezone_name=zone.key)
        minutes = parsed.get("estimated_total_effort_minutes")
        target = _clean_target(parsed.get("title") or "")
        if minutes and target:
            item, _ = match_target(target, items, kinds=("TASK",), statuses=("ACTIVE", "DRAFT"))
            return _action("LOG_PROGRESS", {"minutes": int(minutes)}, item=item, target=target)
    for pattern, command, kinds, statuses in (
        (COMPLETE, "COMPLETE_OBLIGATION", ("TASK", "REMINDER"), open_),
        (ARCHIVE, "ARCHIVE_OBLIGATION", ("TASK",), None),
        (CANCEL, "CANCEL_OBLIGATION", ("TASK", "EVENT", "REMINDER"), open_),
        (COMPLETE_SUFFIX, "COMPLETE_OBLIGATION", ("TASK", "REMINDER"), open_),
        (ARCHIVE_SUFFIX, "ARCHIVE_OBLIGATION", ("TASK",), None),
    ):
        hit = pattern.match(raw)
        if hit:
            target = _clean_target(hit.group("target"))
            item, _ = match_target(target, items, kinds=kinds, statuses=statuses)
            if item is None and pattern in (COMPLETE_SUFFIX, ARCHIVE_SUFFIX):
                continue  # "купить хлеб готово"? only a known item makes a suffix a command
            return _action(command, {}, item=item, target=target, confirm=True)
    moved = MOVE.match(raw)
    if moved:
        parsed = parse_task(moved.group("rest"), now=now, timezone_name=zone.key)
        when = _moment(parsed)
        target = _clean_target(parsed.get("title") or moved.group("rest"))
        item, _ = match_target(target, items, statuses=open_)
        payload: dict[str, Any] = {}
        if when is not None:
            payload = {"when": when.astimezone(ZoneInfo("UTC")).isoformat(),
                       "keep_time": not _CLOCK_WORDS.search(moved.group("rest"))}
        return _action("RESCHEDULE", payload, item=item, target=target, unresolved=[] if when else ["when"])
    snoozed = SNOOZE.match(raw)
    if snoozed:
        rest = snoozed.group("rest")
        parsed = parse_task(rest, now=now, timezone_name=zone.key)
        until = _moment(parsed)
        if until is None and parsed.get("estimated_total_effort_minutes"):
            until = now + timedelta(minutes=int(parsed["estimated_total_effort_minutes"]))  # "отложи на час"
        target = _clean_target(parsed.get("title") or rest)
        item, _ = match_target(target, items, kinds=("TASK", "REMINDER"), statuses=open_)
        if item is None or until is None:
            return None  # "напомни про встречу с деканом завтра": a new reminder
        return _action("SNOOZE", {"until": until.astimezone(ZoneInfo("UTC")).isoformat()}, item=item, target=target)
    return None
